File: solutions/JavaScript/sqrtx_binary_search.js
function sqrtxBinarySearch(x) {
    if (x < 2) return x;

    let left = 2, right = Math.floor(x / 2);
    while (left <= right) {
        const mid = Math.floor((left + right) / 2);
        if (mid * mid === x) {
            return mid;
        } else if (mid * mid < x) {
            left = mid + 1;
        } else {
            right = mid - 1;
        }
    }

    return right;
}

console.log(sqrtxBinarySearch(4));  // 2
console.log(sqrtxBinarySearch(8));  // 2

File: solutions/Java/sqrtx_binary_search.java
public class SqrtxBinarySearch {
    public static int sqrtxBinarySearch(int x) {
        if (x < 2) return x;

        long left = 2, right = x / 2;
        while (left <= right) {
            long mid = (left + right) / 2;
            if (mid * mid == x) {
                return (int) mid;
            } else if (mid * mid < x) {
                left = mid + 1;
            } else {
                right = mid - 1;
            }
        }

        return (int) right;
    }

    public static void main(String[] args) {
        System.out.println(sqrtxBinarySearch(4));  // 2
        System.out.println(sqrtxBinarySearch(8));  // 2
    }
}
